Reformat installation docs to be more readable (#894)
google/knative-gcp: docs/install/pubsub-service-account.md

Pub/Sub. Herein, we show the steps needed to configure such Pub/Sub enabled
Service Account.

## Prerequisites

1. Create a
    [Google Cloud project](https://cloud.google.com/resource-manager/docs/creating-managing-projects)
    and install the `gcloud` CLI and run `gcloud auth login`. This sample will
    use a mix of `gcloud` and `kubectl` commands. The rest of the sample assumes
    that you've set the `$PROJECT_ID` environment variable to your Google Cloud
    project id, and also set your project ID as default using
    `gcloud config set project $PROJECT_ID`.

1. Enable the `Cloud Pub/Sub API` on your project:

    ```shell
    gcloud services enable pubsub.googleapis.com
    ```

## Create a [Google Cloud Service Account](https://console.cloud.google.com/iam-admin/serviceaccounts/project) to interact with Pub/Sub

In general, we would just need permissions to
receive messages (`roles/pubsub.subscriber`). However, in the case of the
`Channel`, we would also need the ability to publish messages
(`roles/pubsub.publisher`).

1. Create a new Service Account named `cre-pubsub` with the following

>     =command:

Reviewer: What's the `=` about?

Author: removed

    ```shell
    gcloud iam service-accounts create cre-pubsub
    ```

2. Give that Service Account the necessary permissions on your project.

    In this example, and for the sake of simplicity, we will just grant
    `roles/pubsub.editor` privileges to the Service Account, which
    encompasses both of the above plus some other permissions. Note that if
    you prefer finer-grained privileges, you can just grant the ones
    mentioned above.

    ```shell
    gcloud projects add-iam-policy-binding $PROJECT_ID \
      --member=serviceAccount:cre-pubsub@$PROJECT_ID.iam.gserviceaccount.com \
      --role roles/pubsub.editor
    ```

## Configure the Authentication Mechanism for GCP

### Option 1: Use Workload Identity

It is the recommended way to access Google Cloud services from within
GKE due to its improved security properties and manageability. For more
information about Workload Identity see
[here](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity).

> 1. Enable Workload Identity. Skip this step if you already enable it in
>   the control plane setup: [Install Knative-GCP](install-knative-gcp.md).

Reviewer 2: ```suggestion
1. Enable Workload Identity. Skip this step if you have already enabled it
  during control plane setup: [Install Knative-GCP](install-knative-gcp.md).
```

    ```shell
    gcloud services enable iamcredentials.googleapis.com
    gcloud beta container clusters update ${CLUSTER_NAME} \
    --identity-namespace=${PROJECT_ID}.svc.id.goog
    ```

> 1. Give `iam.serviceAccountAdmin` role to your control plane's Google
>   Cloud Service Account. You can skip this step if you grant
>   `roles/owner` privileges in the control plane setup:
>   [Install Knative-GCP](install-knative-gcp.md).

Reviewer 2: ```suggestion
1. Give `iam.serviceAccountAdmin` role to your control plane's Google
  Cloud Service Account. You can skip this step if you granted
  `roles/owner` privileges during control plane setup:
  [Install Knative-GCP](install-knative-gcp.md).
```

    ```shell
    gcloud projects add-iam-policy-binding $PROJECT_ID \
    --member=serviceAccount:cloud-run-events@$PROJECT_ID.iam.gserviceaccount.com \
    --role roles/iam.serviceAccountAdmin
    ```

1. Update `spec.googleServiceAccount` with the
  [Google Cloud Service Account](https://console.cloud.google.com/iam-admin/serviceaccounts/project)
  you just created when creating resources. Check
  [example](https://github.com/google/knative-gcp/tree/master/docs/examples)
  of each resource for detail.

    Generally, updating `spec.googleServiceAccount` will trigger the
    controller to enable Workload Identity for sources in the data plane.
    The following steps are handled by the controller:

>     1. Create a Kubernetes Service Account and add an `OwnerReference` to
>      it. The `OwnerReference` is referencing to the source.

Reviewer 2: ```suggestion
    1. Create a Kubernetes Service Account with an `OwnerReference` of
     the source.
```

    1. Bind the Kubernetes Service Account with Google Cloud Service
     Account, this will add `role/iam.workloadIdentityUser` to the
     Google Cloud Service Account. The scope of this role is only for

>      this specific Google Cloud Service Account. It equals to this

Reviewer 2: ```suggestion
     this specific Google Cloud Service Account. It is equivalent to this
```

     command:

        ```shell
        gcloud iam service-accounts add-iam-policy-binding \
        --role roles/iam.workloadIdentityUser \
        --member serviceAccount:$PROJECT_ID.svc.id.goog[namespace/Kubernetes-service-account] \
        GCP-service-account@$PROJECT_ID.iam.gserviceaccount.com
        ```

    1. Annotate the Kubernetes Service Account with
     `iam.gke.io/gcp-service-account=GCP-service-account@PROJECT_ID.iam.gserviceaccount.com`

    For every namespace, the Kubernetes Service Account and the Google

>     Cloud Service Account will have one to one mapping relation. If

Reviewer 2: ```suggestion
    Cloud Service Account will have a one to one mapping. If
```

    multiple source instances use the same Google Cloud Service Account,
    they will use the same Kubernetes Service Account as well, and the

>     kubernetes Service Account will have multiple `OwnerReference`s. If

Reviewer 2: ```suggestion
    Kubernetes Service Account will have multiple `OwnerReference`s. If
```

    there is no source instance using a Kubernetes Service Account, this
    Kubernetes Service Account will be deleted, and its binding to the

>     Google Cloud Service Account will be deleted as well.

Reviewer 2: ```suggestion
    all sources using a Kubernetes Service Account are deleted then that Kubernetes Service Account, and its corresponding Google Cloud Service account, will be deleted.
